File: backend/app/utils/validators.py
import re

from fastapi import HTTPException, status

YOUTUBE_URL_PATTERN = re.compile(
    r"^(https?://)?(www\.)?(youtube\.com/(shorts/|watch\?v=)|youtu\.be/)[\w-]{6,}",
    re.IGNORECASE,
)


def validate_youtube_url(url: str) -> None:
    if not YOUTUBE_URL_PATTERN.search(url.strip()):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Please provide a valid YouTube Shorts/Video URL.",
        )

File: frontend/lib/validators.ts
import { z } from "zod";

const youtubeUrlRegex =
  /^(https?:\/\/)?(www\.)?(youtube\.com\/(shorts\/|watch\?v=)|youtu\.be\/)[\w-]{6,}/i;

export const extractSchema = z.object({
  url: z
    .string()
    .min(1, "YouTube URL is required")
    .regex(youtubeUrlRegex, "Enter a valid YouTube Shorts/Video URL"),
});

export type ExtractFormValues = z.infer<typeof extractSchema>;
